server/src/main/java/org/candlepin/dto/api/v1/ActivationKeyDTO.java

```diff
@@ -0,0 +1,872 @@
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Adds the given product ID to this activation key DTO.
+     *
+     * @param
+     *  productId the product ID to add to this activation key DTO.
+     *
+     * @throws IllegalArgumentException
+     *  if the productId is null or empty
+     *
+     * @return
+     *  true if this product ID was not already contained in this activation key DTO.
+     */
+    @JsonIgnore
+    public boolean addProductId(String productId) {
+        if (productId == null || productId.isEmpty()) {
+            throw new IllegalArgumentException("productId is null or empty");
+        }
+
+        if (this.products == null) {
+            this.products = new HashSet<String>();
+        }
+
+        return this.products.add(productId);
+    }
+
+    /**
+     * Retrieves a view of the content overrides associated with the activation key represented by this DTO.
+     * If the content overrides have not yet been defined, this method returns null.
+     * <p></p>
+     * Note that the collection returned by this method is a view of the collection backing this
+     * set of content overrides. Elements cannot be added to the collection, but elements may be removed.
+     * Changes made to the collection will be reflected by this content overrides data instance.
+     *
+     * @return
+     *  the content overrides associated with this key,
+     *  or null if the content overrideshave not yet been defined
+     */
+    public Set<ActivationKeyContentOverrideDTO> getContentOverrides() {
+        return this.contentOverrides != null ?
+            new SetView<ActivationKeyContentOverrideDTO>(this.contentOverrides) : null;
+    }
+
+    /**
+     * Sets the content overrides of the activation key represented by this DTO.
+     *
+     * @param contentOverrides
+     *  A collection of activation key content override DTOs to attach to this DTO,
+     *  or null to clear the content
+     *
+     * @throws IllegalArgumentException
+     *  if the collection contains null or incomplete content overrides DTOs
+     *
+     * @return
+     *  a reference to this DTO
+     */
+    public ActivationKeyDTO setContentOverrides(Set<ActivationKeyContentOverrideDTO> contentOverrides) {
+        if (contentOverrides != null) {
+            if (this.contentOverrides == null) {
+                this.contentOverrides = new HashSet<ActivationKeyContentOverrideDTO>();
+            }
+            else {
+                this.contentOverrides.clear();
+            }
+
+            for (ActivationKeyContentOverrideDTO dto : contentOverrides) {
+                if (dto == null ||
+                    dto.getContentLabel() == null || dto.getContentLabel().isEmpty() ||
+                    dto.getName() == null || dto.getName().isEmpty() ||
+                    dto.getValue() == null || dto.getValue().isEmpty()) {
+                    throw new IllegalArgumentException("collection contains null " +
```

Comment:
Something I'm just noticing now: We do these checks here to ensure we don't set bad content overrides, but we don't do the same checks in addContentOverride. Immutability aside, we should be consistent with these checks. We should make sure all the places where content overrides can be added are making the same sanity checks.